File: typescript/clase-05/final/main.ts
import Anthropic from "@anthropic-ai/sdk";
import type { MessageParam } from "@anthropic-ai/sdk/resources/messages";
import { createInterface } from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";
import { readFile, writeFile } from "node:fs/promises";

export {};

const MODEL = "claude-sonnet-4-6";
const HISTORY_PATH = "history.json";

async function loadHistory(): Promise<MessageParam[]> {
  try {
    return JSON.parse(await readFile(HISTORY_PATH, "utf8")) as MessageParam[];
  } catch {
    return [];
  }
}

async function saveHistory(messages: MessageParam[]): Promise<void> {
  await writeFile(HISTORY_PATH, JSON.stringify(messages, null, 2));
}

async function streamClaudeResponse(client: Anthropic, messages: MessageParam[]): Promise<string> {
  let assistantText = "";
  const stream = client.messages.stream({ model: MODEL, max_tokens: 700, messages: messages.slice(-12) });
  stream.on("text", (text) => {
    assistantText += text;
    process.stdout.write(text);
  });
  await stream.finalMessage();
  process.stdout.write("\n");
  return assistantText;
}

const apiKey = process.env.ANTHROPIC_API_KEY;
if (!apiKey) throw new Error("Define ANTHROPIC_API_KEY.");

const client = new Anthropic({ apiKey });
const rl = createInterface({ input, output });
let messages = await loadHistory();
console.log("Chatbot listo. Comandos: /salir para terminar, /reset para borrar historial.");

while (true) {
  const userInput = (await rl.question("\nTú: ")).trim();
  if (userInput === "/salir") break;
  if (userInput === "/reset") {
    messages = [];
    await saveHistory(messages);
    console.log("Historial reiniciado.");
    continue;
  }
  if (!userInput) continue;

  messages.push({ role: "user", content: userInput });
  process.stdout.write("Claude: ");
  try {
    const assistantText = await streamClaudeResponse(client, messages);
    messages.push({ role: "assistant", content: assistantText });
    await saveHistory(messages);
  } catch (error) {
    messages.pop();
    console.log(`Error llamando a Claude: ${error instanceof Error ? error.message : "desconocido"}`);
  }
}

rl.close();

File: python/clase-05/final/main.py
"""Chatbot de terminal con historial persistente, comandos, errores y streaming."""

from __future__ import annotations

import json
import os
from pathlib import Path
from anthropic import Anthropic

MODEL = "claude-sonnet-4-6"
HISTORY_PATH = Path("history.json")


def load_history() -> list[dict[str, str]]:
    if HISTORY_PATH.exists():
        return json.loads(HISTORY_PATH.read_text(encoding="utf-8"))
    return []


def save_history(messages: list[dict[str, str]]) -> None:
    HISTORY_PATH.write_text(json.dumps(messages, indent=2, ensure_ascii=False), encoding="utf-8")


def stream_claude_response(client: Anthropic, messages: list[dict[str, str]]) -> str:
    assistant_text = ""
    with client.messages.stream(model=MODEL, max_tokens=700, messages=messages[-12:]) as stream:
        for text in stream.text_stream:
            print(text, end="", flush=True)
            assistant_text += text
    print()
    return assistant_text


def main() -> None:
    api_key = os.getenv("ANTHROPIC_API_KEY")
    if not api_key:
        raise RuntimeError("Define ANTHROPIC_API_KEY.")

    client = Anthropic(api_key=api_key)
    messages = load_history()
    print("Chatbot listo. Comandos: /salir para terminar, /reset para borrar historial.")

    while True:
        user_input = input("\nTú: ").strip()

        if user_input == "/salir":
            break
        if user_input == "/reset":
            messages = []
            save_history(messages)
            print("Historial reiniciado.")
            continue
        if not user_input:
            continue

        messages.append({"role": "user", "content": user_input})
        print("Claude: ", end="", flush=True)

        try:
            assistant_text = stream_claude_response(client, messages)
        except Exception as error:
            messages.pop()
            print(f"Error llamando a Claude: {error}")
            continue

        messages.append({"role": "assistant", "content": assistant_text})
        save_history(messages)


if __name__ == "__main__":
    main()
